In [2]:
import time
import datetime
import Hazard
from Utils.NetworkUtils import *
from Utils.Plot import *
from DynamicNetwork import DynamicNetwork
from HazardMLE import *

# Used to generate dynamic network
start_date = int(time.mktime(datetime.datetime.strptime("08/16/2016", "%m/%d/%Y").timetuple()))
file = "network_data/TheGoodPlace_sample.graphml"   # The good place network
WEEK_IN_SECOND = 7 * 24 * 60 * 60                   # Week in second
HAZARD_BETA = [0.1, 0.1, 0.1]

g = get_graphml(file)
g = sample(g, 10/len(g))      # Get a small sample of network (10 nodes)   
graph_info(g)

Graph nodes: 10, edges: 6


Choose 10 nodes, contain 6 edges.

In [3]:
g = DynamicNetwork(g)   
# Hazard model simulation, using params [beta1, beta2, beta3] = [0.3, 0.3 ,0.3]
model = Hazard.Hazard(g, start_date, WEEK_IN_SECOND, HAZARD_BETA)        
ref_result, fake_data = model.hazard()
print("Adopted users every week: {}, {} steps.".format(ref_result, len(ref_result)))

Adoption Possibility 0.10000, got 0.54856, Not Adopted
Adoption Possibility 0.10000, got 0.50914, Not Adopted
Adoption Possibility 0.10000, got 0.44059, Not Adopted
Adoption Possibility 0.10000, got 0.62114, Not Adopted
Adoption Possibility 0.10000, got 0.93373, Not Adopted
Adoption Possibility 0.10000, got 0.12592, Not Adopted
Adoption Possibility 0.10000, got 0.70829, Not Adopted
Adoption Possibility 0.10000, got 0.12115, Not Adopted
Adoption Possibility 0.10000, got 0.98818, Not Adopted
Adoption Possibility 0.10000, got 0.02407, Adopted
Adoption Possibility 0.10000, got 0.16146, Not Adopted
Adoption Possibility 0.10000, got 0.82785, Not Adopted
Adoption Possibility 0.10000, got 0.61918, Not Adopted
Adoption Possibility 0.10000, got 0.96228, Not Adopted
Adoption Possibility 0.10000, got 0.99208, Not Adopted
Adoption Possibility 0.10000, got 0.93292, Not Adopted
Adoption Possibility 0.10000, got 0.99534, Not Adopted
Adoption Possibility 0.10000, got 0.28994, Not Adopted
Adoption Possi

Adoption Possibility 0.10000, got 0.92597, Not Adopted
Adoption Possibility 0.10000, got 0.03216, Adopted
Adoption Possibility 0.10000, got 0.50541, Not Adopted
Adoption Possibility 0.10000, got 0.35814, Not Adopted
Adoption Possibility 0.10000, got 0.11866, Not Adopted
Adoption Possibility 0.10000, got 0.32942, Not Adopted
Adoption Possibility 0.10000, got 0.81001, Not Adopted
Adoption Possibility 0.10000, got 0.33137, Not Adopted
Adoption Possibility 0.10000, got 0.10694, Not Adopted
Adoption Possibility 0.10000, got 0.16468, Not Adopted
Adoption Possibility 0.10000, got 0.82182, Not Adopted
Adoption Possibility 0.10000, got 0.15994, Not Adopted
Adoption Possibility 0.10000, got 0.07059, Adopted
Adoption Possibility 0.10000, got 0.98185, Not Adopted
Adoption Possibility 0.10000, got 0.31795, Not Adopted
Adoption Possibility 0.10000, got 0.58232, Not Adopted
Adoption Possibility 0.10000, got 0.61264, Not Adopted
Adoption Possibility 0.10000, got 0.67554, Not Adopted
Adoption Possibility 0.10000, got 0.53482, Not Adopted
Adoption Possibility 0.22698, got 0.19975, Adopted
Adoption Possibility 0.10000, got 0.09957, Adopted
Adoption Possibility 0.10000, got 0.08538, Adopted
Adoption Possibility 0.10000, got 0.62262, Not Adopted
Adoption Possibility 0.10000, got 0.87912, Not Adopted
Adoption Possibility 0.10000, got 0.35363, Not Adopted
Adoption Possibility 0.10000, got 0.87550, Not Adopted
Adoption Possibility 0.10000, got 0.92511, Not Adopted
Adoption Possibility 0.10000, got 0.04973, Adopted
Adoption Possibility 0.10000, got 0.41916, Not Adopted
Adoption Possibility 0.10000, got 0.82506, Not Adopted
Adoption Possibility 0.10000, got 0.22832, Not Adopted
Adoption Possibility 0.10000, got 0.60443, Not Adopted
Adoption Possibility 0.10000, got 0.67641, Not Adopted
Adoption Possibility 0.10000, got 0.53070, Not Adopted
Adoption Possibility 0.10000, got 0.75400, Not Adopted
Adoption Possibility 0.10000, got 0.07552, Adopted
Adoption Possibility 0.10000, got 0.66673, Not Adopted
Adoption Possibility 0.10000, got 0.68594, Not Adopted
Adoption Possibility 0.10000, got 0.58834, Not Adopted
Adoption Possibility 0.10000, got 0.96956, Not Adopted
Adoption Possibility 0.10000, got 0.23182, Not Adopted
Adoption Possibility 0.27160, got 0.88343, Not Adopted
Adoption Possibility 0.10000, got 0.30798, Not Adopted
Adoption Possibility 0.10000, got 0.91165, Not Adopted
Adoption Possibility 0.18161, got 0.70517, Not Adopted
Adoption Possibility 0.10000, got 0.94217, Not Adopted
Adoption Possibility 0.10000, got 0.25040, Not Adopted
Adoption Possibility 0.26304, got 0.98134, Not Adopted
Adoption Possibility 0.10000, got 0.53514, Not Adopted
Adoption Possibility 0.10000, got 0.53800, Not Adopted
Adoption Possibility 0.14498, got 0.91067, Not Adopted
Adoption Possibility 0.10000, got 0.03857, Adopted
Adoption Possibility 0.10000, got 0.26498, Not Adopted
Adoption Possibility 0.17635, got 0.87011, Not Adopted
Adoption Possibility 0.10000, got 0.63626, Not Adopted
Adoption Possibility 0.13834, got 0.51082, Not Adopted
Adoption Possibility 0.10000, got 0.01778, Adopted
Adoption Possibility 0.22863, got 0.79931, Not Adopted
Adoption Possibility 0.26575, got 0.29369, Not Adopted
Adoption Possibility 0.15961, got 0.01905, Adopted
Adopted users every week: [1, 2, 5, 6, 7, 7, 7, 7, 7, 7, 8, 8, 9, 9, 10], 15 steps.

In [4]:
def fake_train_data(fake_data):
    from pandas import DataFrame, Series
    train_data_exog = []
    train_data_endog = []
    
    # For demo only begin
    for k, i in fake_data.items():
        train_data_exog.append([k[0], k[1]] + [1] + list(i[1:]) + [i[0]])
        train_data_endog.append(i[0])
    train_data_exog.sort(key=lambda i: (i[0], i[1]))
    train_data_exog = DataFrame(
        train_data_exog, 
        columns=["NODEID", "SECONDS", "CONSTANT", "ADOPTED_NEIGHBORS", "SENTIMENT", "ADOPTION"])
    print(train_data_exog.to_string())
    # For demo only end
    #
    
    train_data_exog = []
    train_data_endog = []
    for k, i in fake_data.items():
        train_data_exog.append([1] + list(i[1:]))
        train_data_endog.append(i[0])
    train_data_exog = DataFrame(train_data_exog, columns=["CONSTANT", "ADOPTED_NEIGHBORS", "SENTIMENT"])
    train_data_endog = Series(train_data_endog, name="ADOPTION")

    return train_data_exog, train_data_endog

# Get exog and endog for mle
exog, endog = fake_train_data(fake_data)

        NODEID     SECONDS  CONSTANT  ADOPTED_NEIGHBORS  SENTIMENT  ADOPTION
0   1184491350  1471330800         1                  0  -0.747110         0
1   1184491350  1471935600         1                  1   0.142406         0
2   1184491350  1472540400         1                  2  -0.528341         0
3   1184491350  1473145200         1                  3  -0.101813         0
4   1184491350  1473750000         1                  4  -0.015771         0
5   1184491350  1474354800         1                  5   0.706360         0
6   1184491350  1474959600         1                  6   0.573778         0
7   1184491350  1475564400         1                  7   0.010195         0
8   1184491350  1476169200         1                  8   0.232355         1
9    124386779  1471330800         1                  0  -0.675931         0
10   124386779  1471935600         1                  1  -0.103657         0
11   124386779  1472540400         1                  2  -0.982007         0

Got data to train MLE.
        NODEID     SECONDS  CONSTANT  ADOPTED_NEIGHBORS  SENTIMENT  ADOPTION
0   1467701994  1471330800         1                  0  -0.868006         0
1   1467701994  1471935600         1                  1   0.670540         0
2   1467701994  1472540400         1                  2   0.269776         1
3     14983059  1471330800         1                  0   0.944534         0
4     14983059  1471935600         1                  1   0.359099         0
5     14983059  1472540400         1                  2   0.943974         0
6     14983059  1473145200         1                  3  -0.412418         1
7   2222550602  1471330800         1                  0  -0.191775         0
8   2222550602  1471935600         1                  1   0.751179         1
9     23249536  1471330800         1                  0   0.755886         0
10    23249536  1471935600         1                  1  -0.317997         0
11    23249536  1472540400         1                  2  -0.134214         0
12    23249536  1473145200         1                  3   0.219756         0
13    23249536  1473750000         1                  4  -0.379094         0
14    23249536  1474354800         1                  5  -0.516655         0
15    23249536  1474959600         1                  6   0.715973         0
16    23249536  1475564400         1                  7  -0.183926         0
17    23249536  1476169200         1                  8   0.630376         0
18    23249536  1476774000         1                  9  -0.550182         0
19    23249536  1477378800         1                 10  -0.236525         0
20    23249536  1477983600         1                 11  -0.616647         0
21    23249536  1478588400         1                 12   0.286287         0
22    23249536  1479193200         1                 13   0.657475         0
23    23249536  1479798000         1                 14  -0.403943         1
24  2457432540  1471330800         1                  0  -0.700006         0
25  2457432540  1471935600         1                  1  -0.066940         0
26  2457432540  1472540400         1                  2  -0.978339         1
27  2715782414  1471330800         1                  0  -0.851808         0
28  2715782414  1471935600         1                  1   0.367569         0
29  2715782414  1472540400         1                  2   0.645550         1
30    38463630  1471330800         1                  0  -0.866894         0
31    38463630  1471935600         1                  1  -0.604140         0
32    38463630  1472540400         1                  2   0.046679         0
33    38463630  1473145200         1                  3   0.596411         0
34    38463630  1473750000         1                  4   0.747701         1
35   439677347  1471330800         1                  0   0.628342         1
36   537417193  1471330800         1                  0   0.791130         0
37   537417193  1471935600         1                  1   0.065054         0
38   537417193  1472540400         1                  2  -0.215020         0
39   537417193  1473145200         1                  3   0.819489         0
40   537417193  1473750000         1                  4  -0.635321         0
41   537417193  1474354800         1                  5   0.964510         0
42   537417193  1474959600         1                  6   0.048261         0
43   537417193  1475564400         1                  7  -0.796117         0
44   537417193  1476169200         1                  8   0.901420         0
45   537417193  1476774000         1                  9   0.166264         0
46   537417193  1477378800         1                 10  -0.403186         0
47   537417193  1477983600         1                 11  -0.331340         0
48   537417193  1478588400         1                 12   0.199875         1
49   616080506  1471330800         1                  0  -0.190076         0
50   616080506  1471935600         1                  1  -0.900545         0
51   616080506  1472540400         1                  2  -0.851069         0
52   616080506  1473145200         1                  3   0.433837         0
53   616080506  1473750000         1                  4   0.693296         0
54   616080506  1474354800         1                  5   0.727239         0
55   616080506  1474959600         1                  6  -0.942758         0
56   616080506  1475564400         1                  7  -0.108206         0
57   616080506  1476169200         1                  8  -0.488052         0
58   616080506  1476774000         1                  9  -0.326715         0
59   616080506  1477378800         1                 10  -0.640373         1

In [5]:
result = HazardModel(exog=exog, endog=endog).fit(method="lbfgs", bounds=[(0.00001, .999), (0.00001, .999), (0.00001, .999)])
print("Beta values {}".format(result.params))

Beta values [  1.00000000e-05   1.00000000e-05   2.54961044e-02]


MLE result:
Beta values [  1.00000000e-05   1.00000000e-05   7.42596219e-02]

In [7]:
def print_loglikelihood(exogs, endogs, params):
    exogs = np.asarray(exogs)
    endogs = np.asarray(endogs)

    log_likelihood = 0

    for exog, endog in zip(exogs, endogs):
        if endog == 1:
            log_likelihood += stats.norm.logcdf(np.dot(exog, params)).sum()
        elif endog == 0:
            log_likelihood += stats.norm.logcdf(-1 * np.dot(exog, params)).sum()
        else:
            assert False, "Shouldn't run into this line"

    print("{}, {}".format(params, log_likelihood))
    
# params = [round(p, 3) for p in result.params] # Reset params to 3 digits after decimal
print("MLE loglikelihood")
print_loglikelihood(exog, endog, result.params)
print("Original loglikelihood")
print_loglikelihood(exog, endog, HAZARD_BETA)

MLE loglikelihood
[  1.00000000e-05   1.00000000e-05   2.54961044e-02], -57.5287507273502
Original loglikelihood
[0.1, 0.1, 0.1], -96.24581429294444


Compare loglikelihood:

MLE loglikelihood
[  1.00000000e-05   1.00000000e-05   2.54961044e-02], -57.5287507273502

Original beta loglikelihood
[0.3, 0.3, 0.3], -257.6248851705792